### PR TITLE
feat: add cli for scaffolding a new challenge

### PR DESCRIPTION
**Pull request by nbaars**
Still work in progress.

Refs: #555

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[package]
```
Comment (commjoen) [inline cli/Cargo.toml:1]: Can you please add this path to the dependabot configuration :)

File: cli/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  challenge  Create a new challenge
+  help       Print this message or the help of the given subcommand(s)
+
+Options:
+  -h, --help  Print help
+```
+
+## Building
+
+First install [Rust](https://www.rust-lang.org/tools/install). Then open a terminal and type:
```
Comment (commjoen) [inline cli/README.md:30]: Maybe we can have a cross compiling and building github action with a download link to its attachments so people can copy it as well?

File: cli/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+When passing command line arguments you need to add them to the run configuration. In IntelliJ go to `Run` -> `Edit Configurations...` and add the arguments to the `Command` field. You need to add `--` before the arguments. For example:
+
+```shell
+run --package challenge-cli --bin challenge-cli -- challenge -d easy -t git ../
+```
+
+## Todo
+
+- Fix templating (not everything is present yet)
+- Add GitHub actions to build binary for the different platforms
```
Comment (commjoen) [inline cli/README.md:50]: ah yes :D this!

File: src/main/resources/challenge.hbs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+{{
+extra_imports
+}}
+
+@Component
+@Order(0)
```
Comment (commjoen) [inline src/main/resources/challenge.hbs:18]: can we make the order also automatede in terms o the number?

File: src/main/resources/challenge.hbs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import org.owasp.wrongsecrets.challenges.Spoiler;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+{{
+extra_imports
+}}
+
```
Comment (commjoen) [inline src/main/resources/challenge.hbs:16]:
```suggestion
/*
* <put your javadoc here describing the challenge in one sentence>
*/
```

File: src/main/resources/challenge.hbs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int difficulty() {
+        return Difficulty.{{difficulty}};
+    }
+
```
Comment (commjoen) [inline src/main/resources/challenge.hbs:56]:
```suggestion
     /**
     * {@inheritDoc}
     * uses {{technology}}.
     */
```

File: src/main/resources/challenge.hbs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        return false;
+    }
+
+    @Override
+    public boolean canRunInCTFMode() {
+        return true;
+    }
+
+    private String getData() {
+        return "<<replace with correct answer>>";
```
Comment (commjoen) [inline src/main/resources/challenge.hbs:73]:
```suggestion
        return "<<replace with correct answer but then obfuscated/encrypted in code>>";
```

File: cli/README.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+On `main.rs` right click and select `Run 'main'`. This will run the CLI in the terminal window of IntelliJ.
+When passing command line arguments you need to add them to the run configuration. In IntelliJ go to `Run` -> `Edit Configurations...` and add the arguments to the `Command` field. You need to add `--` before the arguments. For example:
+
+```shell
+run --package challenge-cli --bin challenge-cli -- challenge -d easy -t git ../
+```
+
+## Todo
+
+- Fix templating (not everything is present yet)
```
Comment (commjoen) [inline cli/README.md:49]: it would be nice to have some basic tests as well i guess :) ? Can you add those please?

**Review by commjoen — CHANGES_REQUESTED** [review]
Wow @nbaars this is a great start!
Thank you so much. I have some first feedback, can you please have a look at it ?

**Review by commjoen — CHANGES_REQUESTED** [review]
Wow @nbaars this is a great start!
Thank you so much. I have some first feedback, can you please have a look at it ?

**Comment by commjoen** [comment]
Elligible for `hacktoberfest-accepted` label if timely finished in october

**Comment by commjoen** [comment]
Hi @nbaars ! Would you like to create a new iteration based on the refactor results please?

**Comment by commjoen** [comment]
Hi @nbaars should we keep this one open to generate the required files :-)?

**Comment by nbaars** [comment]
I'm going to rewrite this with [Spring Boot CLI](https://docs.spring.io/spring-boot/docs/current/reference/html/cli.html). Much cleaner approach.

**Pull request closed without merging**